### ISSUE TITLE
Add hexagonal architecture graph mojo

### ISSUE THREAD

**Issue by javathought** [binout/livingdoc-maven-plugin#3]
Add a mojo to generate graph with hexagonal architecture view

**Comment by binout** [comment]
Good idea,
I think of using dot-diagram (https://github.com/cyriux/dot-diagram) but it's not published in maven repo ... I ask to Cyrille but no answer for the moment

**Comment by javathought** [comment]
Yes, I saw the dependence. It will be the easier if he do so.

**Comment by javathought** [comment]
No news from Cyrille ?

**Comment by binout** [comment]
No ... a solution is to fork its project and create a "io.github.livingdocumenation.dot-diagram" ?
I'm almost sure he 'll be ok ;-)

**Comment by javathought** [comment]
Ok, I'll asked a teammate that has worked with him some years ago if he had news from him.

**Comment by binout** [comment]
I have opened an issue : https://github.com/cyriux/dot-diagram/issues/5

**Comment by binout** [comment]
dot-diagram is now in the organization :-)

**Comment by binout** [comment]
We can create a feature branch to work on this feature ?

**Comment by javathought** [comment]
I've initiated the feature branch : feature/#3_hexa_arch_graph_mojo 

Some hard initialization should be replaced by plugin parameters.

**Comment by javathought** [comment]
It's a harder part to migrate : the base code make use of class path loaded by running test (`ClassPath.from(Thread.currentThread().getContextClassLoader());`). 
So not trivial to make it a maven plugin, and we should have a to get another ClassLoader

From maven wiki (https://cwiki.apache.org/confluence/display/MAVEN/Maven+3.x+Class+Loading)
> Plugins that need to load classes from the compile/runtime/test class path of a project need to create a custom URLClassLoader in combination with the mojo annotation @requiresDependencyResolution.

One detailed solution, to try : http://stackoverflow.com/questions/871708/maven-plugin-cant-load-class

**Comment by javathought** [comment]
It's on the way 👌 : it works.
Remaining tasks : 
 - Some cleanup
 - add parameters 
 - discuss on how to identify Core Domain

@binout @cyriux :
Actual code is based on package name containing "domain" : solutions
- we can continue with the method, replacing hard coded value by parameter
- we can replace by an annotation (thus adding a new module to embed annotations in target project)

**Comment by cyriux** [comment]
I'd start with the simplest option (package name with "domain" in the name), then evolve based on users feedbacks. 

An important point is to explain the naming convention. For example the mojo could dump a file naming-conventions.txt with the current naming convention described for humans in plain English. 

Le 29 nov. 2016 à 00:30, javathought <notifications@github.com> a écrit :

It's on the way 👌 : it works.
Remaining tasks :

Some cleanup
add parameters
discuss on how to identify Core Domain
@binout @cyriux :
Actual code is based on package name containing "domain" : solutions

we can continue with the method, replacing hard coded value by parameter
we can replace by an annotation (thus adding a new module to embed annotations in target project)

**Comment by javathought** [comment]
I'll keep this simpliest option for this 1st implementation, with the name defined by a mojo parameter.

**Comment by binout** [comment]
@javathought do you really need to instantiate classes to analyse code ?

You can use https://github.com/paul-hammant/qdox to scan java source files such as GlossaryMojo (https://github.com/LivingDocumentation/livingdoc-maven-plugin/blob/develop/livingdoc-maven-plugin/src/main/java/io/github/livingdocumentation/GlossaryMojo.java#L45).

**Comment by cyriux** [comment]
Le 29 nov. 2016 à 18:26, Benoît Prioux <notifications@github.com> a écrit :
Qdox is great indeed!
@javathought do you really need to instantiate classes to analyse code ?

You can use https://github.com/paul-hammant/qdox to scan java source files such as GlossaryMojo (https://github.com/LivingDocumentation/livingdoc-maven-plugin/blob/develop/livingdoc-maven-plugin/src/main/java/io/github/livingdocumentation/GlossaryMojo.java#L45).

**Comment by javathought** [comment]
It's the current implementation in original code. So I'm migrating as is. We can refactor with Qdox. 

I also have a use case of running the plugin on a module with a dependency on another : one module is the core domain, which is a dependance of others one. So with runtime a module, I should get my dependency in the diagram without adding sources to the mojo. 

It can then also add external project classes in the diagram if wanted. This is not in my personal use case.

What about migrating as mojo first with less rewrite as possible and keeping refactor before release of 1.0 ?

Le 29 nov. 2016 à 18:26, Benoît Prioux @binout  a écrit :
@javathought do you really need to instantiate classes to analyse code ?

**Comment by javathought** [comment]
closed by PR #15
